php/libraries/SubprojectController.class.inc

```diff
@@ -0,0 +1,252 @@
+     *
+     * @param \LORIS\Subproject $subproject A subproject object without the ID
+     *
+     * @return \LORIS\Subproject The created subproject with it's new ID
+     */
+    public function createSubproject(\LORIS\Subproject $subproject)
+    : \LORIS\Subproject
+    {
+        if (!empty($subproject->getID())) {
+            throw new \LorisException(
+                "An ID can not be supplied when creating a new subproject, the ID
+                will be generated by the database."
+            );
+        }
+        $subprojectExists = $this->subprojectExists($subproject->getTitle());
+
+        if (!$subprojectExists) {
+            throw new \LorisException(
+                "A subproject with the same title ".
+                $subproject->getTitle()." already exists"
+            );
+        }
+
+        // Update subproject table
+        $this->database->insert(
+            'subproject',
+            [
+                'title'             => $subproject->getTitle(),
+                'useEDC'            => is_null($subproject->getUseEDC())
+                    ?: ($subproject->getUseEDC() ? 1 : 0),
+                'WindowDifference'  => $subproject->getWindowDifference(),
+                'RecruitmentTarget' => $subproject->getRecruitmentTarget(),
+            ]
+        );
+
+        return $this->getSubprojectByTitle($subproject->getTitle());
+    }
+
+    /**
+     * Update an existing subproject in the database.
+     *
+     * @param \LORIS\Subproject $subproject The subproject object to update
+     *
+     * @return \LORIS\Subproject The updated subproject
+     */
+    public function updateSubproject(\LORIS\Subproject $subproject)
+    : \LORIS\Subproject
+    {
+        // Update subproject table
+        $this->database->update(
+            'subproject',
+            [
+                'title'             => $subproject->getTitle(),
+                'useEDC'            => is_null($subproject->getUseEDC())
+                    ?: ($subproject->getUseEDC() ? 1 : 0),
+                'WindowDifference'  => $subproject->getWindowDifference(),
+                'RecruitmentTarget' => $subproject->getRecruitmentTarget(),
+            ],
+            ['title'  => $subproject->getTitle()]
+        );
+
+        return $this->getSubprojectByTitle($subproject->getTitle());
+    }
+
+    /**
+     * Check if a subproject already exists in the DB based on the title
+     *
+     * @param string $subprojectTitle Title of the subproject
+     *
+     * @return bool
+     */
+    public function subprojectExists(string $subprojectTitle)
+    {
+        $subproject = $this->database->pselectOne(
+            "SELECT SubprojectID FROM subproject WHERE title=:title",
+            ["title"=>$subprojectTitle]
+        );
+
+        if (empty($subproject)) {
+            return false;
+        }
+        return true;
+    }
+
+    //TODO: loadProjects
+    //TODO: loadVisits(Project)
```

Comment:
I don't think a SubprojectController should produce Projects nor Visits. I think ProjectController and VisitController should produce them based on a given Subproject. Inversely, subprojectController should produce subprojects based on given visits and/or projects.

```
getSubprojectFromProject(\Project $p): \Traversable
{
    $subprojects =  $this->database->pselect(
        "SELECT 
            sp.*
            from project_subproject_rel psr
            LEFT JOIN subproject sp ON (psr.SubprojectID = sp.SubprojectID)
            WHERE psr.ProjectID = :v_projectid",
        ['v_projectid' => $p->getID()]
    );
    ... if PDO::FETCH_CLASS is used then ...
    yield from $subprojects;
}

getSubprojectFromVisit(\Visit $v): Subproject
{
   ...
}
```

Reply:
completely agree